# Data Preprocessing

This notebook prepares the project data for exploratory analysis, visualization, and modeling. The goal is to create a clean match-level dataset for FIFA World Cup match outcome prediction while avoiding data leakage.

## Dataset usage decisions

- **Primary modeling dataset:** `results.csv`
  - Contains one row per international match.
  - Includes match date, home team, away team, scores, tournament, city, country, and neutral-site status.
  - Serves as the base table because the model predicts outcomes at the match level.

- **Main external feature dataset:** `eloratings.csv`
  - Contains historical Elo ratings by team and date.
  - Elo ratings are merged onto the match results data to create pre-match team-strength features.
  - Main Elo-based features: `home_elo`, `away_elo`, and `elo_diff`.

- **Datasets reviewed but not merged into the first modeling dataset:**
  - `matches.csv`, `teams.csv`, `tournaments.csv`, and `tournament_standings.csv` contain useful World Cup metadata for EDA or later feature engineering.
  - Some columns in these files describe post-match or post-tournament outcomes, such as match result, score margin, tournament winner, final standing, extra time, and penalty shootout information. These would not be known before kickoff and could introduce data leakage if used as predictors.
  - `goalscorers.csv` and `shootouts.csv` are event-level datasets that describe events during or after a match. They may be useful later if aggregated carefully using only information available before each match date, but they are excluded from the first modeling dataset.

- **Output files created by this notebook:**
  - `matches_preprocessed_readable.csv`: readable match-level dataset for EDA and visualization.
  - `matches_modeling_base.csv`: modeling dataset with selected pre-match features before encoding/scaling.
  - `X_train_processed.csv`, `X_test_processed.csv`, `y_train.csv`, and `y_test.csv`: encoded/scaled train/test files for modeling.
  - `feature_names.csv` and `label_mapping.csv`: metadata to help interpret processed model inputs and encoded target labels.


## Imports and Data Loading

The notebook loads all raw datasets so that their structures can be reviewed, but the first modeling pipeline uses `results.csv` and `eloratings.csv` as the main inputs.


In [1]:
import pandas as pd
import numpy as np
from pathlib import Path

# Use ../data when running from a notebooks/ folder; fall back to ./data when running from the repo root.
DATA_DIR = Path("../data")
if not (DATA_DIR / "results.csv").exists():
    DATA_DIR = Path("./data")

PROCESSED_DIR = DATA_DIR / "processed"
PROCESSED_DIR.mkdir(parents=True, exist_ok=True)

print("Using data directory:", DATA_DIR.resolve())
print("Processed outputs will be saved to:", PROCESSED_DIR.resolve())


Using data directory: /Users/oliviadarby/DATASCI_207/ucb_mids_207_world_cup_prediction_model/data
Processed outputs will be saved to: /Users/oliviadarby/DATASCI_207/ucb_mids_207_world_cup_prediction_model/data/processed


In [2]:
# Fjelstul World Cup Database files
matches = pd.read_csv(DATA_DIR / "matches.csv")
teams = pd.read_csv(DATA_DIR / "teams.csv")
tournaments = pd.read_csv(DATA_DIR / "tournaments.csv")
standings = pd.read_csv(DATA_DIR / "tournament_standings.csv")

# International Football Results files
results = pd.read_csv(DATA_DIR / "results.csv")
goalscorers = pd.read_csv(DATA_DIR / "goalscorers.csv")
shootouts = pd.read_csv(DATA_DIR / "shootouts.csv")

# Elo ratings file
eloratings = pd.read_csv(DATA_DIR / "eloratings.csv")


In [3]:
# Structure check
dataframes = {
    "matches": matches,
    "teams": teams,
    "tournaments": tournaments,
    "standings": standings,
    "results": results,
    "shootouts": shootouts,
    "goalscorers": goalscorers,
    "eloratings": eloratings
}

for name, df in dataframes.items():
    print(f"{name}: {df.shape}")
    display(df.head())


matches: (1248, 37)


,key_id,tournament_id,tournament_name,match_id,match_name,stage_name,group_name,group_stage,knockout_stage,replayed,...,away_team_score_margin,extra_time,penalty_shootout,score_penalties,home_team_score_penalties,away_team_score_penalties,result,home_team_win,away_team_win,draw
0,1,WC-1930,1930 FIFA Men's World Cup,M-1930-01,France vs Mexico,group stage,Group 1,1,0,0,...,-3,0,0,0-0,0,0,home team win,1,0,0
1,2,WC-1930,1930 FIFA Men's World Cup,M-1930-02,United States vs Belgium,group stage,Group 4,1,0,0,...,-3,0,0,0-0,0,0,home team win,1,0,0
2,3,WC-1930,1930 FIFA Men's World Cup,M-1930-03,Yugoslavia vs Brazil,group stage,Group 2,1,0,0,...,-1,0,0,0-0,0,0,home team win,1,0,0
3,4,WC-1930,1930 FIFA Men's World Cup,M-1930-04,Romania vs Peru,group stage,Group 3,1,0,0,...,-2,0,0,0-0,0,0,home team win,1,0,0
4,5,WC-1930,1930 FIFA Men's World Cup,M-1930-05,Argentina vs France,group stage,Group 1,1,0,0,...,-1,0,0,0-0,0,0,home team win,1,0,0


teams: (88, 14)


,key_id,team_id,team_name,team_code,mens_team,womens_team,federation_name,region_name,confederation_id,confederation_name,confederation_code,mens_team_wikipedia_link,womens_team_wikipedia_link,federation_wikipedia_link
0,1,T-01,Algeria,DZA,1,0,Algerian Football Federation,Africa,CF-2,Confederation of African Football,CAF,https://en.wikipedia.org/wiki/Algeria_national...,not applicable,https://en.wikipedia.org/wiki/Algerian_Footbal...
1,2,T-02,Angola,AGO,1,0,Angolan Football Federation,Africa,CF-2,Confederation of African Football,CAF,https://en.wikipedia.org/wiki/Angola_national_...,not applicable,https://en.wikipedia.org/wiki/Angolan_Football...
2,3,T-03,Argentina,ARG,1,1,Argentine Football Association,South America,CF-4,South American Football Confederation,CONMEBOL,https://en.wikipedia.org/wiki/Argentina_nation...,https://en.wikipedia.org/wiki/Argentina_women'...,https://en.wikipedia.org/wiki/Argentine_Footba...
3,4,T-04,Australia,AUS,1,1,Football Australia,Oceania,CF-1,Asian Football Confederation,AFC,https://en.wikipedia.org/wiki/Australia_men%27...,https://en.wikipedia.org/wiki/Australia_women'...,https://en.wikipedia.org/wiki/Football_Australia
4,5,T-05,Austria,AUT,1,0,Austrian Football Association,Europe,CF-6,Union of European Football Associations,UEFA,https://en.wikipedia.org/wiki/Austria_national...,not applicable,https://en.wikipedia.org/wiki/Austrian_Footbal...


tournaments: (30, 18)


,key_id,tournament_id,tournament_name,year,start_date,end_date,host_country,winner,host_won,count_teams,group_stage,second_group_stage,final_round,round_of_16,quarter_finals,semi_finals,third_place_match,final
0,1,WC-1930,1930 FIFA Men's World Cup,1930,1930-07-13,1930-07-30,Uruguay,Uruguay,1,13,1,0,0,0,0,1,0,1
1,2,WC-1934,1934 FIFA Men's World Cup,1934,1934-05-27,1934-06-10,Italy,Italy,1,16,0,0,0,1,1,1,1,1
2,3,WC-1938,1938 FIFA Men's World Cup,1938,1938-06-04,1938-06-19,France,Italy,0,15,0,0,0,1,1,1,1,1
3,4,WC-1950,1950 FIFA Men's World Cup,1950,1950-06-24,1950-07-16,Brazil,Uruguay,0,13,1,0,1,0,0,0,0,0
4,5,WC-1954,1954 FIFA Men's World Cup,1954,1954-06-16,1954-07-04,Switzerland,West Germany,0,16,1,0,0,0,1,1,1,1


standings: (120, 7)


,key_id,tournament_id,tournament_name,position,team_id,team_name,team_code
0,1,WC-1930,1930 FIFA Men's World Cup,1,T-84,Uruguay,URY
1,2,WC-1930,1930 FIFA Men's World Cup,2,T-03,Argentina,ARG
2,3,WC-1930,1930 FIFA Men's World Cup,3,T-83,United States,USA
3,4,WC-1930,1930 FIFA Men's World Cup,4,T-87,Yugoslavia,YUG
4,5,WC-1934,1934 FIFA Men's World Cup,1,T-41,Italy,ITA


results: (49477, 9)


,date,home_team,away_team,home_score,away_score,tournament,city,country,neutral
0,1872-11-30,Scotland,England,0.0,0.0,Friendly,Glasgow,Scotland,False
1,1873-03-08,England,Scotland,4.0,2.0,Friendly,London,England,False
2,1874-03-07,Scotland,England,2.0,1.0,Friendly,Glasgow,Scotland,False
3,1875-03-06,England,Scotland,2.0,2.0,Friendly,London,England,False
4,1876-03-04,Scotland,England,3.0,0.0,Friendly,Glasgow,Scotland,False


shootouts: (678, 5)


,date,home_team,away_team,winner,first_shooter
0,1967-08-22,India,Taiwan,Taiwan,NaN
1,1971-11-14,South Korea,Vietnam Republic,South Korea,NaN
2,1972-05-07,South Korea,Iraq,Iraq,NaN
3,1972-05-17,Thailand,South Korea,South Korea,NaN
4,1972-05-19,Thailand,Cambodia,Thailand,NaN


goalscorers: (47606, 8)


,date,home_team,away_team,team,scorer,minute,own_goal,penalty
0,1916-07-02,Chile,Uruguay,Uruguay,José Piendibene,44.0,False,False
1,1916-07-02,Chile,Uruguay,Uruguay,Isabelino Gradín,55.0,False,False
2,1916-07-02,Chile,Uruguay,Uruguay,Isabelino Gradín,70.0,False,False
3,1916-07-02,Chile,Uruguay,Uruguay,José Piendibene,75.0,False,False
4,1916-07-06,Argentina,Chile,Argentina,Alberto Ohaco,2.0,False,False


eloratings: (6678, 4)


,date,team,rating,change
0,1872-11-30,England,2003.0,3
1,1872-11-30,Scotland,1997.0,-3
2,1873-03-08,England,2014.0,11
3,1873-03-08,Scotland,1986.0,-11
4,1874-03-07,England,2006.0,-8


## Missing Value Handling and Basic Cleaning

The main supervised-learning target requires completed matches. Rows without final home or away scores are removed because their outcomes are not known. Dates are also converted to datetime format so the Elo merge and time-based split work correctly.


In [4]:
# Convert date columns.
# Elo dates use mixed formats, so format="mixed" is used.
results["date"] = pd.to_datetime(results["date"], errors="coerce")
eloratings["date"] = pd.to_datetime(eloratings["date"], format="mixed", errors="coerce")

print("Missing results dates:", results["date"].isna().sum())
print("Missing Elo dates:", eloratings["date"].isna().sum())


Missing results dates: 0
Missing Elo dates: 0


In [5]:
# Drop matches without final scores because they cannot be used for supervised learning.
missing_scores = results[results["home_score"].isna() | results["away_score"].isna()]
print("Rows with missing home or away score:", missing_scores.shape[0])

results = results.dropna(subset=["home_score", "away_score"]).copy()

# Convert scores to integers after removing missing score rows.
results["home_score"] = results["home_score"].astype(int)
results["away_score"] = results["away_score"].astype(int)

print("Results shape after score cleaning:", results.shape)
print("Duplicate rows in results:", results.duplicated().sum())
results.head()


Rows with missing home or away score: 70
Results shape after score cleaning: (49407, 9)
Duplicate rows in results: 0


,date,home_team,away_team,home_score,away_score,tournament,city,country,neutral
0,1872-11-30,Scotland,England,0,0,Friendly,Glasgow,Scotland,False
1,1873-03-08,England,Scotland,4,2,Friendly,London,England,False
2,1874-03-07,Scotland,England,2,1,Friendly,Glasgow,Scotland,False
3,1875-03-06,England,Scotland,2,2,Friendly,London,England,False
4,1876-03-04,Scotland,England,3,0,Friendly,Glasgow,Scotland,False


## Target Variable and Initial Feature Engineering

The target variable is `match_result`, with three classes: `home_win`, `away_win`, and `draw`. Score-based columns are created for EDA and validation, but they are later excluded from the modeling feature set because they are only known after the match.


In [6]:
def get_match_result(row):
    if row["home_score"] > row["away_score"]:
        return "home_win"
    elif row["home_score"] < row["away_score"]:
        return "away_win"
    else:
        return "draw"

results["match_result"] = results.apply(get_match_result, axis=1)
results["goal_diff"] = results["home_score"] - results["away_score"]
results["abs_goal_diff"] = results["goal_diff"].abs()
results["total_goals"] = results["home_score"] + results["away_score"]
results["year"] = results["date"].dt.year
results["is_world_cup"] = results["tournament"].str.contains("World Cup", case=False, na=False)

print(results["match_result"].value_counts())
results[[
    "date", "year", "home_team", "away_team", "home_score", "away_score",
    "match_result", "goal_diff", "abs_goal_diff", "total_goals", "tournament", "is_world_cup"
]].head(10)


match_result
home_win    24214
away_win    13960
draw        11233
Name: count, dtype: int64


,date,year,home_team,away_team,home_score,away_score,match_result,goal_diff,abs_goal_diff,total_goals,tournament,is_world_cup
0,1872-11-30,1872,Scotland,England,0,0,draw,0,0,0,Friendly,False
1,1873-03-08,1873,England,Scotland,4,2,home_win,2,2,6,Friendly,False
2,1874-03-07,1874,Scotland,England,2,1,home_win,1,1,3,Friendly,False
3,1875-03-06,1875,England,Scotland,2,2,draw,0,0,4,Friendly,False
4,1876-03-04,1876,Scotland,England,3,0,home_win,3,3,3,Friendly,False
5,1876-03-25,1876,Scotland,Wales,4,0,home_win,4,4,4,Friendly,False
6,1877-03-03,1877,England,Scotland,1,3,away_win,-2,2,4,Friendly,False
7,1877-03-05,1877,Wales,Scotland,0,2,away_win,-2,2,2,Friendly,False
8,1878-03-02,1878,Scotland,England,7,2,home_win,5,5,9,Friendly,False
9,1878-03-23,1878,Scotland,Wales,9,0,home_win,9,9,9,Friendly,False


## Outlier Detection and Treatment

Scoreline outliers are checked using absolute goal differential. Extreme results are inspected rather than automatically removed. Because these appear to be legitimate historical matches rather than data-entry errors, they are retained.


In [7]:
results["abs_goal_diff"].describe()


count    49407.000000
mean         1.717469
std          1.793860
min          0.000000
25%          1.000000
50%          1.000000
75%          2.000000
max         31.000000
Name: abs_goal_diff, dtype: float64

In [8]:
extreme_scorelines = results[results["abs_goal_diff"] >= 20].copy()

print("Number of matches with absolute goal differential >= 20:", extreme_scorelines.shape[0])

extreme_scorelines[[
    "date", "home_team", "away_team", "home_score", "away_score", "tournament", "abs_goal_diff"
]].sort_values("abs_goal_diff", ascending=False)


Number of matches with absolute goal differential >= 20: 13


,date,home_team,away_team,home_score,away_score,tournament,abs_goal_diff
25425,2001-04-11,Australia,American Samoa,31,0,FIFA World Cup qualification,31
8551,1971-09-13,Tahiti,Cook Islands,30,0,South Pacific Games,30
11916,1979-08-30,Fiji,Kiribati,24,0,South Pacific Games,24
25422,2001-04-09,Australia,Tonga,22,0,FIFA World Cup qualification,22
37061,2013-06-24,Provence,Tibet,22,0,"International Tournament of Peoples, Cultures ...",22
6580,1966-04-03,Libya,Oman,21,0,Arab Cup,21
29045,2005-03-11,Guam,North Korea,0,21,EAFF Championship,21
37059,2013-06-23,Quebec,Tibet,21,0,"International Tournament of Peoples, Cultures ...",21
15922,1987-12-15,American Samoa,Papua New Guinea,0,20,South Pacific Games,20
24181,2000-02-14,Kuwait,Bhutan,20,0,AFC Asian Cup qualification,20


## Team Name Standardization

Team names must be standardized before merging match results with Elo ratings. Manual mappings are applied only for clear naming differences between the two datasets. Remaining unmatched teams are reviewed; most are regional, non-FIFA, or representative teams and are not manually forced into a national-team mapping.


In [9]:
# Clean hidden/non-breaking spaces in Elo team names.
eloratings["team"] = (
    eloratings["team"]
    .str.replace("\xa0", " ", regex=False)
    .str.strip()
)

# Candidate mappings for common naming differences.
POSSIBLE_TEAM_NAME_MAPPING = {
    "Czech Republic": "Czechia",
    "Republic of Ireland": "Ireland",
    "Chinese Taipei": "Taiwan",
    "USA": "United States",
    "United States of America": "United States",
    "DR Congo": "Democratic Republic of Congo",
    "Democratic Republic of the Congo": "Democratic Republic of Congo",
    "German DR": "East Germany",
    "North Macedonia": "Macedonia",
    "FYR Macedonia": "Macedonia",
    "Eswatini": "Swaziland",
    "Myanmar": "Burma",
    "Macau": "Macao",
    "Timor-Leste": "East Timor",
    "Cabo Verde": "Cape Verde",
    "Côte d'Ivoire": "Ivory Coast",
    "Cote d'Ivoire": "Ivory Coast",
    "Türkiye": "Turkey",
    "Curaçao": "Curacao",
    "São Tomé and Príncipe": "Sao Tome and Principe",
    "São Tomé and Principe": "Sao Tome and Principe",
    "Sao Tome and Príncipe": "Sao Tome and Principe",
    "St. Kitts and Nevis": "Saint Kitts and Nevis",
    "St Kitts and Nevis": "Saint Kitts and Nevis",
    "St. Lucia": "Saint Lucia",
    "St Lucia": "Saint Lucia",
    "St. Vincent and the Grenadines": "Saint Vincent and the Grenadines",
    "St Vincent and the Grenadines": "Saint Vincent and the Grenadines",
    "United States Virgin Islands": "US Virgin Islands",
    "Kyrgyz Republic": "Kyrgyzstan",
    "Brunei Darussalam": "Brunei",
    "Lao": "Laos",
    "Lao PDR": "Laos"
}

# Keep only mappings whose target names actually appear in the Elo file.
elo_team_names = set(eloratings["team"].dropna().unique())
TEAM_NAME_MAPPING = {
    old: new
    for old, new in POSSIBLE_TEAM_NAME_MAPPING.items()
    if new in elo_team_names
}

print("Mappings used:")
TEAM_NAME_MAPPING


Mappings used:


{'Czech Republic': 'Czechia',
 'Republic of Ireland': 'Ireland',
 'Chinese Taipei': 'Taiwan',
 'USA': 'United States',
 'United States of America': 'United States',
 'DR Congo': 'Democratic Republic of Congo',
 'Democratic Republic of the Congo': 'Democratic Republic of Congo',
 'German DR': 'East Germany',
 'North Macedonia': 'Macedonia',
 'FYR Macedonia': 'Macedonia',
 'Eswatini': 'Swaziland',
 'Myanmar': 'Burma',
 'Macau': 'Macao',
 'Timor-Leste': 'East Timor',
 'Cabo Verde': 'Cape Verde',
 "Côte d'Ivoire": 'Ivory Coast',
 "Cote d'Ivoire": 'Ivory Coast',
 'Türkiye': 'Turkey',
 'São Tomé and Príncipe': 'Sao Tome and Principe',
 'São Tomé and Principe': 'Sao Tome and Principe',
 'Sao Tome and Príncipe': 'Sao Tome and Principe',
 'St. Kitts and Nevis': 'Saint Kitts and Nevis',
 'St Kitts and Nevis': 'Saint Kitts and Nevis',
 'St. Lucia': 'Saint Lucia',
 'St Lucia': 'Saint Lucia',
 'St. Vincent and the Grenadines': 'Saint Vincent and the Grenadines',
 'St Vincent and the Grenadines': 'S

In [10]:
results["home_team_clean"] = results["home_team"].replace(TEAM_NAME_MAPPING)
results["away_team_clean"] = results["away_team"].replace(TEAM_NAME_MAPPING)
eloratings["team_clean"] = eloratings["team"].replace(TEAM_NAME_MAPPING)

results_teams = set(results["home_team_clean"]).union(set(results["away_team_clean"]))
elo_teams = set(eloratings["team_clean"])

unmatched_results_teams = sorted(results_teams - elo_teams)
print("Number of teams in results not found in Elo:", len(unmatched_results_teams))
unmatched_results_teams[:50]


Number of teams in results not found in Elo: 97


['Abkhazia',
 'Alderney',
 'Ambazonia',
 'American Samoa',
 'Andalusia',
 'Arameans Suryoye',
 'Artsakh',
 'Asturias',
 'Aymara',
 'Barawa',
 'Basque Country',
 'Biafra',
 'Brittany',
 'Canary Islands',
 'Cascadia',
 'Catalonia',
 'Central Spain',
 'Chameria',
 'Chechnya',
 'Cilento',
 'Corsica',
 'County of Nice',
 'Crimea',
 'Darfur',
 'Donetsk PR',
 'Délvidék',
 'East Turkestan',
 'Elba Island',
 'Ellan Vannin',
 'Felvidék',
 'Franconia',
 'Frøya',
 'Galicia',
 'Gotland',
 'Gozo',
 'Guernsey',
 'Găgăuzia',
 'Hitra',
 'Hmong',
 'Iraqi Kurdistan',
 'Isle of Man',
 'Isle of Wight',
 'Jersey',
 'Kabylia',
 'Kernow',
 'Kárpátalja',
 'Luhansk PR',
 'Madrid',
 'Manchukuo',
 'Mapuche']

In [11]:
missing_elo_team_set = set(unmatched_results_teams)

rows_with_unmatched_team = results[
    results["home_team_clean"].isin(missing_elo_team_set) |
    results["away_team_clean"].isin(missing_elo_team_set)
]

print("Rows involving unmatched teams:", rows_with_unmatched_team.shape[0])
print("Percent of results data:", round(100 * rows_with_unmatched_team.shape[0] / results.shape[0], 2))

unmatched_home_counts = results.loc[
    results["home_team_clean"].isin(missing_elo_team_set),
    "home_team_clean"
].value_counts()

unmatched_away_counts = results.loc[
    results["away_team_clean"].isin(missing_elo_team_set),
    "away_team_clean"
].value_counts()

unmatched_counts = unmatched_home_counts.add(unmatched_away_counts, fill_value=0).sort_values(ascending=False)
unmatched_counts.head(30)


Rows involving unmatched teams: 1611
Percent of results data: 3.26


Guernsey                     240.0
Jersey                       235.0
Vietnam Republic             195.0
Alderney                     135.0
Réunion                      124.0
Ynys Môn                      67.0
Basque Country                64.0
Isle of Man                   58.0
Shetland                      58.0
American Samoa                55.0
Åland Islands                 51.0
Isle of Wight                 48.0
Catalonia                     48.0
Padania                       45.0
Frøya                         37.0
Saare County                  35.0
Western Isles                 34.0
Occitania                     33.0
Abkhazia                      33.0
Western Australia             32.0
Gotland                       30.0
Sápmi                         29.0
Orkney                        28.0
Tamil Eelam                   28.0
Iraqi Kurdistan               27.0
Székely Land                  27.0
Raetia                        26.0
Yemen DPR                     25.0
Menorca             

## Merge Home and Away Elo Ratings

Elo ratings are merged using the most recent rating available **before** each match date. `allow_exact_matches=False` is used because ratings recorded on the same date as a match may already include the result of that match. This avoids data leakage.


In [12]:
# Sort dataframes before merge_asof.
results_sorted = results.sort_values("date").copy()
elo_sorted = eloratings.sort_values("date").copy()

# Merge most recent available home-team Elo rating before the match date.
home_merged = pd.merge_asof(
    results_sorted,
    elo_sorted[["date", "team_clean", "rating"]].sort_values("date"),
    left_on="date",
    right_on="date",
    left_by="home_team_clean",
    right_by="team_clean",
    direction="backward",
    allow_exact_matches=False
)

home_merged = home_merged.rename(columns={"rating": "home_elo"})
home_merged = home_merged.drop(columns=["team_clean"])

# Merge most recent available away-team Elo rating before the match date.
model_df = pd.merge_asof(
    home_merged.sort_values("date"),
    elo_sorted[["date", "team_clean", "rating"]].sort_values("date"),
    left_on="date",
    right_on="date",
    left_by="away_team_clean",
    right_by="team_clean",
    direction="backward",
    allow_exact_matches=False
)

model_df = model_df.rename(columns={"rating": "away_elo"})
model_df = model_df.drop(columns=["team_clean"])

# Create Elo differential.
model_df["elo_diff"] = model_df["home_elo"] - model_df["away_elo"]

model_df[["date", "home_team", "away_team", "home_elo", "away_elo", "elo_diff"]].head()


,date,home_team,away_team,home_elo,away_elo,elo_diff
0,1872-11-30,Scotland,England,NaN,NaN,NaN
1,1873-03-08,England,Scotland,2003.0,1997.0,6.0
2,1874-03-07,Scotland,England,1986.0,2014.0,-28.0
3,1875-03-06,England,Scotland,2006.0,1994.0,12.0
4,1876-03-04,Scotland,England,1997.0,2003.0,-6.0


In [13]:
missing_elo_rows = model_df[
    model_df["home_elo"].isna() | model_df["away_elo"].isna()
]

print("Rows after Elo merge:", model_df.shape[0])
print("Rows missing at least one Elo rating:", missing_elo_rows.shape[0])
print("Percent missing Elo:", round(100 * missing_elo_rows.shape[0] / model_df.shape[0], 2))
print("\nMissing Elo values by column:")
print(model_df[["home_elo", "away_elo", "elo_diff"]].isna().sum())


Rows after Elo merge: 49407
Rows missing at least one Elo rating: 7447
Percent missing Elo: 15.07

Missing Elo values by column:
home_elo    4690
away_elo    5053
elo_diff    7447
dtype: int64


In [14]:
# Check missing Elo coverage for World Cup matches.
world_cup_df = model_df[model_df["is_world_cup"]].copy()
wc_missing_elo = world_cup_df[
    world_cup_df["home_elo"].isna() | world_cup_df["away_elo"].isna()
]

print("World Cup rows:", world_cup_df.shape[0])
print("World Cup rows missing at least one Elo:", wc_missing_elo.shape[0])
print("Percent World Cup missing Elo:", round(100 * wc_missing_elo.shape[0] / world_cup_df.shape[0], 2))


World Cup rows: 9798
World Cup rows missing at least one Elo: 773
Percent World Cup missing Elo: 7.89


In [15]:
# Compare possible year restrictions by Elo coverage.
coverage_summary = []

for cutoff in [1950, 1970, 1990, 1994, 2000, 2010]:
    subset = model_df[model_df["year"] >= cutoff]
    missing = subset[
        subset["home_elo"].isna() | subset["away_elo"].isna()
    ]
    coverage_summary.append({
        "cutoff_year": cutoff,
        "rows": subset.shape[0],
        "missing_elo_rows": missing.shape[0],
        "missing_elo_percent": round(100 * missing.shape[0] / subset.shape[0], 2),
        "usable_elo_rows": subset.dropna(subset=["home_elo", "away_elo", "elo_diff"]).shape[0]
    })

coverage_summary_df = pd.DataFrame(coverage_summary)
coverage_summary_df


,cutoff_year,rows,missing_elo_rows,missing_elo_percent,usable_elo_rows
0,1950,46069,5765,12.51,40304
1,1970,41447,4096,9.88,37351
2,1990,32289,2653,8.22,29636
3,1994,29904,2295,7.67,27609
4,2000,25345,1703,6.72,23642
5,2010,15819,906,5.73,14913


## Year Restriction

The full match dataset dates back to 1872. The main modeling dataset is restricted to matches from **1994 onward** to better reflect the modern international soccer environment while keeping a large sample size. This cutoff also improves Elo coverage compared with the full historical dataset.


In [16]:
START_YEAR = 1994

model_df_modern = model_df[model_df["year"] >= START_YEAR].copy()

missing_elo_modern = model_df_modern[
    model_df_modern["home_elo"].isna() | model_df_modern["away_elo"].isna()
]

model_elo_df = model_df_modern.dropna(
    subset=["home_elo", "away_elo", "elo_diff"]
).copy()

print("Original dataset shape:", model_df.shape)
print("Modern dataset shape:", model_df_modern.shape)
print("Rows missing at least one Elo in modern data:", missing_elo_modern.shape[0])
print("Percent missing Elo in modern data:", round(100 * missing_elo_modern.shape[0] / model_df_modern.shape[0], 2))
print("Final Elo modeling dataset rows:", model_elo_df.shape[0])


Original dataset shape: (49407, 20)
Modern dataset shape: (29904, 20)
Rows missing at least one Elo in modern data: 2295
Percent missing Elo in modern data: 7.67
Final Elo modeling dataset rows: 27609


## Save Readable EDA Dataset

This dataset keeps human-readable columns and retains rows with missing Elo values. It is intended for EDA and visualization, not final model fitting.


In [17]:
readable_cols = [
    "date",
    "year",
    "home_team",
    "away_team",
    "home_team_clean",
    "away_team_clean",
    "home_score",
    "away_score",
    "match_result",
    "tournament",
    "city",
    "country",
    "neutral",
    "is_world_cup",
    "home_elo",
    "away_elo",
    "elo_diff",
    "goal_diff",
    "abs_goal_diff",
    "total_goals"
]

eda_df = model_df_modern[readable_cols].copy()
eda_df.to_csv(PROCESSED_DIR / "matches_preprocessed_readable.csv", index=False)

print("Readable EDA dataset shape:", eda_df.shape)
print("Saved to:", PROCESSED_DIR / "matches_preprocessed_readable.csv")


Readable EDA dataset shape: (29904, 20)
Saved to: ../data/processed/matches_preprocessed_readable.csv


## Feature Selection

Features are selected based on whether they would be known before kickoff. Elo ratings, Elo differential, tournament context, neutral-site status, and World Cup indicator are retained as predictors. Score-based columns such as home score, away score, goal differential, absolute goal differential, and total goals are excluded from the modeling feature set because they directly reveal information about the match outcome.


In [18]:
target_col = "match_result"

numeric_features = [
    "home_elo",
    "away_elo",
    "elo_diff"
]

categorical_features = [
    "tournament",
    "neutral",
    "is_world_cup"
]

# date and year are retained for temporal splitting and traceability, but they are not used as model predictors below.
modeling_cols = numeric_features + categorical_features + [target_col, "date", "year"]
modeling_base_df = model_elo_df[modeling_cols].copy()

modeling_base_df.to_csv(PROCESSED_DIR / "matches_modeling_base.csv", index=False)

print("Modeling base dataset shape:", modeling_base_df.shape)
print("Saved modeling base dataset:", PROCESSED_DIR / "matches_modeling_base.csv")
modeling_base_df.head()


Modeling base dataset shape: (27609, 9)
Saved modeling base dataset: ../data/processed/matches_modeling_base.csv


,home_elo,away_elo,elo_diff,tournament,neutral,is_world_cup,match_result,date,year
19503,1248.0,1363.0,-115.0,Friendly,False,False,draw,1994-01-02,1994
19504,1622.0,1663.0,-41.0,Friendly,False,False,home_win,1994-01-02,1994
19505,1501.0,1324.0,177.0,Friendly,False,False,draw,1994-01-05,1994
19506,1077.0,1501.0,-424.0,Friendly,False,False,away_win,1994-01-09,1994
19507,1354.0,1622.0,-268.0,Friendly,False,False,draw,1994-01-11,1994


## Temporal Train/Test Split

A time-based split is used so that later matches are treated as future observations. This is more appropriate for prediction than randomly mixing historical and recent matches across train and test sets.


In [19]:
train_df = modeling_base_df[modeling_base_df["year"] < 2022].copy()
test_df = modeling_base_df[modeling_base_df["year"] >= 2022].copy()

print("Train shape:", train_df.shape)
print("Test shape:", test_df.shape)
print("Train years:", train_df["year"].min(), "-", train_df["year"].max())
print("Test years:", test_df["year"].min(), "-", test_df["year"].max())

print("\nTrain target distribution:")
display(train_df[target_col].value_counts(normalize=True).rename("proportion"))

print("\nTest target distribution:")
display(test_df[target_col].value_counts(normalize=True).rename("proportion"))


Train shape: (23192, 9)
Test shape: (4417, 9)
Train years: 1994 - 2021
Test years: 2022 - 2026

Train target distribution:


match_result
home_win    0.484132
away_win    0.277208
draw        0.238660
Name: proportion, dtype: float64


Test target distribution:


match_result
home_win    0.477926
away_win    0.289789
draw        0.232284
Name: proportion, dtype: float64

## Encoding and Scaling

Categorical predictors are one-hot encoded so they can be used in machine learning models. Numeric Elo-based features are standardized using the training data only, then the same transformation is applied to the test data. This prevents information from the test set from influencing preprocessing.


In [20]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, StandardScaler, LabelEncoder

X_train = train_df[numeric_features + categorical_features].copy()
X_test = test_df[numeric_features + categorical_features].copy()

y_train = train_df[target_col].copy()
y_test = test_df[target_col].copy()

preprocessor = ColumnTransformer(
    transformers=[
        ("num", StandardScaler(), numeric_features),
        ("cat", OneHotEncoder(handle_unknown="ignore"), categorical_features)
    ],
    remainder="drop"
)

X_train_processed = preprocessor.fit_transform(X_train)
X_test_processed = preprocessor.transform(X_test)

label_encoder = LabelEncoder()
y_train_encoded = label_encoder.fit_transform(y_train)
y_test_encoded = label_encoder.transform(y_test)

print("X_train processed shape:", X_train_processed.shape)
print("X_test processed shape:", X_test_processed.shape)
print("Target classes:", list(label_encoder.classes_))


X_train processed shape: (23192, 102)
X_test processed shape: (4417, 102)
Target classes: ['away_win', 'draw', 'home_win']


In [21]:
# Convert processed arrays back to dataframes for easier saving and inspection.
feature_names = preprocessor.get_feature_names_out()

X_train_processed_df = pd.DataFrame(
    X_train_processed.toarray() if hasattr(X_train_processed, "toarray") else X_train_processed,
    columns=feature_names,
    index=train_df.index
)

X_test_processed_df = pd.DataFrame(
    X_test_processed.toarray() if hasattr(X_test_processed, "toarray") else X_test_processed,
    columns=feature_names,
    index=test_df.index
)

y_train_df = pd.DataFrame(
    {"match_result": y_train.values, "match_result_encoded": y_train_encoded},
    index=train_df.index
)

y_test_df = pd.DataFrame(
    {"match_result": y_test.values, "match_result_encoded": y_test_encoded},
    index=test_df.index
)

X_train_processed_df.head()


,num__home_elo,num__away_elo,num__elo_diff,cat__tournament_ABCS Tournament,cat__tournament_AFC Asian Cup,cat__tournament_AFC Asian Cup qualification,cat__tournament_AFC Challenge Cup,cat__tournament_AFC Challenge Cup qualification,cat__tournament_AFC Solidarity Cup,cat__tournament_AFF Championship,...,cat__tournament_USA Cup,cat__tournament_United Arab Emirates Friendship Tournament,cat__tournament_Unity Cup,cat__tournament_VFF Cup,cat__tournament_WAFF Championship,cat__tournament_Windward Islands Tournament,cat__neutral_False,cat__neutral_True,cat__is_world_cup_False,cat__is_world_cup_True
19503,-1.003883,-0.511272,-0.488182,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
19504,0.344179,0.552091,-0.219132,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
19505,-0.091959,-0.649509,0.573474,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
19506,-1.620243,-0.022125,-1.611647,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0
19507,-0.621812,0.406765,-1.044461,0.0,0.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,1.0,0.0,1.0,0.0


In [22]:
# Save final train/test files.
X_train_processed_df.to_csv(PROCESSED_DIR / "X_train_processed.csv", index=False)
X_test_processed_df.to_csv(PROCESSED_DIR / "X_test_processed.csv", index=False)
y_train_df.to_csv(PROCESSED_DIR / "y_train.csv", index=False)
y_test_df.to_csv(PROCESSED_DIR / "y_test.csv", index=False)

# Save metadata for interpretability.
feature_names_df = pd.DataFrame({"feature_name": feature_names})
label_mapping_df = pd.DataFrame({
    "match_result": label_encoder.classes_,
    "match_result_encoded": range(len(label_encoder.classes_))
})

feature_names_df.to_csv(PROCESSED_DIR / "feature_names.csv", index=False)
label_mapping_df.to_csv(PROCESSED_DIR / "label_mapping.csv", index=False)

print("Saved processed train/test files and metadata.")


Saved processed train/test files and metadata.


## Preprocessing Summary

- Matches without final scores were removed because they cannot be used for supervised learning.
- Scoreline outliers were inspected using absolute goal differential and retained because they appeared to be legitimate historical results.
- Team names were standardized before merging with Elo ratings.
- Elo ratings were merged using the most recent rating available before the match date to avoid data leakage.
- Rows with missing Elo values were retained in the readable EDA dataset but removed from the Elo-based modeling dataset because Elo differential is a core predictor.
- The main modeling dataset was restricted to matches from 1994 onward to better reflect modern international soccer while preserving a large sample size.
- Only pre-match features were retained for modeling.
- Score-based variables were excluded from predictors because they directly reveal the match outcome.
- Categorical variables were one-hot encoded.
- Numeric Elo features were standardized using the training data only.
- A time-based train/test split was used so that later matches are evaluated as future observations.
